### PR TITLE
Re-factor macro-flag handling, fix Table-returning paths, only allow DataFrames 0.22 and above

### PR DESCRIPTION
**Pull request by pdeffebach**
This PR implements the following syntax

```
julia> df = DataFrame(a = [1, 2, 3], b = [4, 5, 6]);

julia> @transform df @astable (x = :a, y = :b)
3×4 DataFrame
 Row │ a      b      x      y     
     │ Int64  Int64  Int64  Int64 
─────┼────────────────────────────
   1 │     1      4      1      4
   2 │     2      5      2      5
   3 │     3      6      3      6
```

It also re-configures the way we treat macro flags so it will be easier to generalize flags in the future. 

I still need to add docs and tests.

### CODE REVIEW

**Comment by bkamins** [comment]
Thank you. Please let me know when it is good for a review.

**Comment by pdeffebach** [comment]
Okay this is ready for a review. 

It's not obvious how useful this is, I guess. It doesn't do the automatic NamedTuple creation with `:x = ...` as in [DFMacros](https://github.com/jkrumbiegel/DataFrameMacros.jl#t-flag-macro-for-automatic-astable). 

But I guess it's less clunky than `cols(AsTable) = ...`.

**Comment by bkamins** [comment]
Is the intended use case of this is something like:
```
df = DataFrame(a=[[1,2], [2,3]])
@select(df, @astable(:a))
```
?
(I am asking because most of the examples it tests seemed a bit artificial to me.

**Comment by pdeffebach** [comment]
That example would work. 

To be clear, there is no benefit from `AsTable` that you can't get with `cols(AsTable) = ...` currently. 

I ended up adding this because I knew I needed to re-factor the table-producing paths before I implement `@select`. If we feel like this isn't a priority, I can change this PR to get rid of `@astable` but keep the good refactoring and the new macro flag logic.

**Comment by pdeffebach** [comment]
I'm going to keep this as a re-factoring PR to keep moving forward on the more essential parts.

**Comment by bkamins** [comment]
I understand that in the future it would be `$(AsTable)` right? In such case do you think there would be a big benefit of `@astable` over it? (and we could save introducing a new pseudo-macro)

CC @nalimilan

**Comment by pdeffebach** [comment]
Okay this PR is ready for a review. 

There are some minor changes to how we handle `@byrow` and `@with` internally, but neither of these affect any user-facing APIs. 

The only thing I need a review for is the function `extract_macro_flags`, and whether the implementation isn't too complex. 

I am going to *not* add docs about `cols(AsTable)` usage right now, in order to move forward with `@subset`.

**Comment by bkamins** [comment]
CI fails

**Comment by pdeffebach** [comment]
Okay tests pass now. 

Basically, before, `@byrow` stuff was handled in two places, the function that makes the `src => fun` part and the more general expression handling function. Now the `@byrow` stuff is handled exclusively in the "expression handling" function.

**Comment by pdeffebach** [comment]
Given this is just about internals and not the API, I will merge after just an approval from @bkamins

**Review by bkamins — APPROVED** [review]
Thank you! (I have not investigated every individual line, but I assume you are pretty sure about this internals refactor?)

**Comment by pdeffebach** [comment]
Yes. At the very least, something similar was required before `@subset` so better be a separate PR for simplicity.

**Pull request merged**
